### PR TITLE
Implement references as objects.

### PR DESCRIPTION
**Pull request by smimram**
Implement references as objects: `ref('a)` is now `(() -> 'a).{set : ('a) -> unit}`. For instance,
```
x = ref(5)
x.set(x() + 2)
x := x() + 2
```
(the `:=` syntax is still supported for now)

### CODE REVIEW

**Comment by smimram** [comment]
I find the syntax proposed above quite heavy actually. Another option would be to implement `ref('a)` as
```
(() -> 'a).{set : ('a) -> unit
```
For instance,
```
x = ref(5)
x.set(x() + 2)
```
This would have in particular the advantage that any reference is directly a valid getter!

One of the aims was to remove the `!` notation on the long term. I think that this would be 

What do you think?

**Comment by toots** [comment]
I think that:

```
(() -> 'a).{set : ('a) -> unit
```
should be the way to go, this is how the settings are already implemented!

How about type safety? In OCaml we have this:
```ocaml
# let f =
    let x = ref [] in
    fun v -> x := v::!x;;
val f : '_a -> unit = <fun>
```
Where the type `'_a` can. only be set once and for all.

**Comment by smimram** [comment]
The _value restriction_ seems to work here (= `'_a` in OCaml): you cand test it with `test/failures/ref.liq`.

File: src/core/lang_source.ml [hunk 1]
```diff
@@ -542,6 +542,7 @@ let iter_sources ?on_reference ~static_analysis_failed f v =
             List.iter (function _, _, Some v -> iter_value v | _ -> ()) proto
         | FFI (proto, _) ->
             List.iter (function _, _, Some v -> iter_value v | _ -> ()) proto
```
Comment (toots) [inline src/core/lang_source.ml:544]: I think we should simply call `on_imprecise` when reaching a `FFI`. We can't check the function's body in this case so this makes sense and will suit our case here.

**Review by toots — APPROVED** [review]
Looking great just one remark.

**Pull request merged**